src/dbnode/network/server/tchannelthrift/errors/errors.go

```diff
@@ -43,14 +44,31 @@ func IsBadRequestError(err *rpc.Error) bool {
 	return err != nil && err.Type == rpc.ErrorType_BAD_REQUEST
 }
 
+// IsResourceExhaustedError returns whether the error is a resource exhausted error.
+func IsResourceExhaustedError(err *rpc.Error) bool {
+	// NB: To maintain better backwards compatibility, Resource Exhausted errors might also be
+	// defined by BAD_REQUEST error type coupled with RESOURCE_EXHAUSTED error code
+	return err != nil &&
+		(err.Type == rpc.ErrorType_RESOURCE_EXHAUSTED ||
+			(err.Type == rpc.ErrorType_BAD_REQUEST && err.Code == rpc.ErrorCode_RESOURCE_EXHAUSTED))
+}
+
 // NewInternalError creates a new internal error
 func NewInternalError(err error) *rpc.Error {
-	return newError(rpc.ErrorType_INTERNAL_ERROR, err)
+	return newError(rpc.ErrorType_INTERNAL_ERROR, rpc.ErrorCode_NONE, err)
 }
 
 // NewBadRequestError creates a new bad request error
 func NewBadRequestError(err error) *rpc.Error {
-	return newError(rpc.ErrorType_BAD_REQUEST, err)
+	return newError(rpc.ErrorType_BAD_REQUEST, rpc.ErrorCode_NONE, err)
+}
+
+// NewResourceExhaustedError creates a new resource exhausted error.
+func NewResourceExhaustedError(err error) *rpc.Error {
+	// NB: To maintain better backwards compatibility, using BAD_REQUEST error type coupled with
+	// RESOURCE_EXHAUSTED error code. After a reasonable amount of time this should be switched to
+	// RESOURCE_EXHAUSTED error code (added after M3 release v1.0.0)
+	return newError(rpc.ErrorType_BAD_REQUEST, rpc.ErrorCode_RESOURCE_EXHAUSTED, err)
```

Comment:
This should introduce the protocol changes in a backwards compatible way. Older clients, which are not aware about `Resource Exhausted` error type, will fallback to current `Bad Request` behavior. On the other hand, newer client would know about `Resource Exhausted` errors, but the server would keep sending `Bad Request` until it's upgraded.